nixos/modules/virtualisation/xen-dom0.nix

```diff
@@ -1,213 +1,925 @@
-    virtualisation.xen.stored =
-      mkOption {
-        type = types.path;
-        description =
-          ''
-            Xen Store daemon to use. Defaults to oxenstored of the xen package.
-          '';
+      bootParams = lib.options.mkOption {
+        default = [ ];
+        example = ''
+          [
+            "iommu=force:true,qinval:true,debug:true"
+            "noreboot=true"
+            "vga=ask"
+          ]
+        '';
+        type = lib.types.listOf lib.types.str;
+        description = ''
+          Xen Command Line parameters passed to Domain 0 at boot time.
+          Note: these are different from `boot.kernelParams`. See
+          the [Xen documentation](https://xenbits.xenproject.org/docs/unstable/misc/xen-command-line.html) for more information.
+        '';
       };
 
-    virtualisation.xen.domains = {
-        extraConfig = mkOption {
-          type = types.lines;
+      domains = {
+        extraConfig = lib.options.mkOption {
+          type = lib.types.lines;
           default = "";
-          description =
-            ''
-              Options defined here will override the defaults for xendomains.
-              The default options can be seen in the file included from
-              /etc/default/xendomains.
-            '';
-          };
+          example = ''
+            XENDOMAINS_SAVE=/persist/xen/save
+            XENDOMAINS_RESTORE=false
+            XENDOMAINS_CREATE_USLEEP=10000000
+          '';
+          description = ''
+            Options defined here will override the defaults for xendomains.
+            The default options can be seen in the file included from
+            /etc/default/xendomains.
+          '';
+        };
       };
+    };
+  };
 
-    virtualisation.xen.trace = mkEnableOption "Xen tracing";
+  ###### implementation
 
-  };
+  config = lib.modules.mkIf cfg.enable {
+    assertions = [
+      {
+        assertion = pkgs.stdenv.isx86_64;
+        message = "Xen is currently not supported on ${pkgs.stdenv.hostPlatform.system}.";
+      }
+      {
+        assertion = cfg.efi.enable -> config.boot.loader.systemd-boot.enable;
+        message = "Xen currently does not support EFI boot on bootloaders that are not systemd-boot.";
+      }
+      {
+        assertion = !config.boot.loader.systemd-boot.bootCounting.enable;
+        message = "Xen currently does not support systemd-boot's boot counting feature.";
+      }
+      {
+        assertion = cfg.dom0Resources.maxMemory >= cfg.dom0Resources.memory;
+        message = ''
+          You have allocated more memory to dom0 than virtualisation.xen.dom0Resources.maxMemory
+          allows for. Please increase the maximum memory limit, or decrease the default memory allocation.
+        '';
+      }
+      {
+        assertion = if cfg.debug then if cfg.trace then true else false else true;
+        message = "Xen's debugging features are enabled, but logging is disabled. This is most likely not what you want.";
+      }
```

Comment:
This is kinda weird, can't we just make it do trace if debug is enabled instead?

Reply:
`debug` without `trace` makes no sense in any situation, so a warning would allow the user to see undefined behaviour. 

I agree that it should be enabled by default though if `debug` is enabled.